components/raftstore/src/store/snap/io.rs

```diff
@@ -295,20 +295,101 @@ where
     }
 }
 
-pub fn apply_sst_cf_file<E>(files: &[&str], db: &E, cf: &str) -> Result<(), Error>
+pub fn apply_sst_cf_files_by_ingest<E>(files: &[&str], db: &E, cf: &str) -> Result<(), Error>
 where
     E: KvEngine,
 {
     if files.len() > 1 {
         info!(
-            "apply_sst_cf_file starts on cf {}. All files {:?}",
+            "apply_sst_cf_files_by_ingest starts on cf {}. All files {:?}",
             cf, files
         );
     }
     box_try!(db.ingest_external_file_cf(cf, files));
     Ok(())
 }
 
+fn apply_sst_cf_file_without_ingest<E, F>(
+    path: &str,
+    db: &E,
+    cf: &str,
+    key_mgr: Option<Arc<DataKeyManager>>,
+    stale_detector: &impl StaleDetector,
+    batch_size: usize,
+    callback: &mut F,
+) -> Result<(), Error>
+where
+    E: KvEngine,
+    F: for<'r> FnMut(&'r [(Vec<u8>, Vec<u8>)]),
+{
+    let sst_reader = E::SstReader::open(path, key_mgr)?;
+    let mut iter = sst_reader.iter(IterOptions::default())?;
+    iter.seek_to_first()?;
+
+    let mut wb = db.write_batch();
+    let mut write_to_db = |batch: &mut Vec<(Vec<u8>, Vec<u8>)>| -> Result<(), EngineError> {
+        batch.iter().try_for_each(|(k, v)| wb.put_cf(cf, k, v))?;
+        wb.write()?;
+        wb.clear();
+        callback(batch);
+        batch.clear();
+        Ok(())
+    };
+
+    // Collect keys to a vec rather than wb so that we can invoke the callback less
+    // times.
+    let mut batch = Vec::with_capacity(1024);
```

Comment:
How about using WriteBatch directly?

Reply:
Not appropriate for this case. 

Just like `apply_plain_file` does, it should notify the cop with `post_apply_plain_kvs_from_snapshot` when each batch are flushed. But `WriteBatch` does not support this with callback.

Or we can add a new function for `WriteBatch` ? Make it can notify the cop by `callback (F: for<'r> FnMut(&'r [(Vec<u8>, Vec<u8>)]))` ?